### PR TITLE
Gas estimation: Overestimate based on (actor, method) tuples

### LINKED ISSUE #8417: Gas Estimation around NV16
While the final details of the FVM gas model are TBD, it seems likely that the cost of messages will increase after the upgrade to nv16. This poses a challenge for messages that are in the mempool at the time of the upgrade: they will have been estimated with the LegacyVM, and therefore are susceptible to running out of gas when they are executed using the FVM. 

We have a few options here:

1. Do nothing. A few epochs worth of messages will fail with Out of Gas, but all messages estimated after the upgrade will be fine. Most critical messages in Lotus have retry mechanisms on failure, so nothing too bad will happen.
2. Start naively overestimating at some period before the upgrade, by applying (say) a 2x factor. The question would be when to start doing this -- too early and messages that land before the upgrade will pay a lot in overestimation, too late and there will be some messages that were already estimated (and so might fail out of gas).
3. Start estimating using the actual FVM at some period before the upgrade. The same question about when to start would apply. This won't be too hard, but is probably more work than it's worth.

### PR DESCRIPTION
**Pull request by Kubuxu**
## Related Issues

Resolves #8417

## Proposed Changes

Have few unique multipleirs apart from the base multipleir

### CODE REVIEW

File: node/impl/full/gas.go [hunk 1]
```diff
@@ -297,6 +297,47 @@ func gasEstimateGasLimit(
...
+				return
+			}
+			act, err := st.GetActor(msg.To)
+			if err != nil {
+				return
+			}
+
+			if builtin.IsStorageMinerActor(act.Code) {
+				switch msgIn.Method {
+				case 5:
```
Comment (arajasek) [inline node/impl/full/gas.go:317]: `case miner.Methods.SubmitWindowedPoSt`? (and so on for others)

File: node/impl/full/gas.go [hunk 2]
```diff
@@ -297,6 +297,47 @@ func gasEstimateGasLimit(
 
 	ret := res.MsgRct.GasUsed
 
+	transitionalMulti := 1.0
+	// Overestimate gas around the upgrade
+	if ts.Height() <= build.UpgradeFVM1Height && (build.UpgradeFVM1Height-ts.Height() <= 20) {
+		transitionalMulti = 2.0
+
+		func() {
```
Comment (arajasek) [inline node/impl/full/gas.go:305]: what do we gain by using an anonymous function here?

File: node/impl/full/gas.go [hunk 3]
```diff
@@ -297,6 +297,47 @@ func gasEstimateGasLimit(
 
 	ret := res.MsgRct.GasUsed
 
+	transitionalMulti := 1.0
+	// Overestimate gas around the upgrade
+	if ts.Height() <= build.UpgradeFVM1Height && (build.UpgradeFVM1Height-ts.Height() <= 20) {
```
Comment (arajasek) [inline node/impl/full/gas.go:302]: How did you pick 20 here?

**Review by arajasek — APPROVED** [review]
Thanks for the PR! I have some qs / suggestions, but good enough to merge!

File: node/impl/full/gas.go (hunk 1 above)
Comment (Kubuxu) [reply node/impl/full/gas.go:317]: This corresponds to what is reported in collected statistics so it is easier to verify if it stays as numbers, also as this is very temporary code there is little to no harm to having raw numbers.

If you feel strongly about it I can add comments with method names.

File: node/impl/full/gas.go (hunk 2 above)
Comment (Kubuxu) [reply node/impl/full/gas.go:305]: It cleans up the error handling.

File: node/impl/full/gas.go (hunk 3 above)
Comment (Kubuxu) [reply node/impl/full/gas.go:302]: This gives us 10 minutes of overestimation, there are few reasons I went with 20 epochs.
I would be comfortable with 5 minutes, if we didn’t have tipset system, possible off-by-ones somewhere so on, and I feel like increasing it to 10min has a low cost.

File: node/impl/full/gas.go (hunk 1 above)
Comment (arajasek) [reply node/impl/full/gas.go:317]: I don't feel strongly about it, no.

if you want me to double check the values against the collected statistics, i can, but i trust you

File: node/impl/full/gas.go (hunk 2 above)
Comment (arajasek) [reply node/impl/full/gas.go:305]: ah, makes sense

File: node/impl/full/gas.go (hunk 3 above)
Comment (arajasek) [reply node/impl/full/gas.go:302]: Seems reasonable -- i don't think there's a "correct" answer here.

**Comment by Fatman13** [comment]
> Do nothing. A few epochs worth of messages will fail with Out of Gas, but all messages estimated after the upgrade will be fine. Most critical messages in Lotus have retry mechanisms on failure, so nothing too bad will happen.

> Start naively overestimating at some period before the upgrade, by applying (say) a 2x factor. The question would be when to start doing this -- too early and messages that land before the upgrade will pay a lot in overestimation, too late and there will be some messages that were already estimated (and so might fail out of gas).

> Start estimating using the actual FVM at some period before the upgrade. The same question about when to start would apply. This won't be too hard, but is probably more work than it's worth.

Is this PR still going for option 2 like the one before?

**Pull request merged**